In [125]:
# importing essential libraries
import pandas as pd

# importing classes and dataset from sklearn library
from sklearn import datasets
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB, MultinomialNB
from sklearn.neighbors import KNeighborsClassifier as knn
from sklearn.model_selection import GridSearchCV

In [126]:
irisdata = datasets.load_iris()
# X = pd.DataFrame(irisdata.data, columns = ["SepalLength", "SepalWidth", "PetalLength", "PetalWidth"])
# Y = pd.DataFrame(irisdata.target, columns = ["Flower"])
# print(X.shape)
# print(Y.shape)

In [127]:
Methods = {
    
    # model
    "SVM" : {
        # implementing model and setting it to Model key
        "Model" : svm.SVC(),
        
        # hyperparameters list
        "HyperParams" : {
            "C" : [1.0, 5.0, 10.0],
            "kernel" : ["linear", "rbf"],
            "gamma" : ["auto", "scale"]
        }
    },
    
    'RandomForestClassification' : {
        "Model" : RandomForestClassifier(max_features = "auto"),
        "HyperParams" : {
            "n_estimators" : [10, 30, 50]
        }
    },
    
    "KNN" : {
        "Model" : knn(),
        "HyperParams" : {
            "n_neighbors" : [4, 5, 6]
        }
    },
    
    "LogisticRegression" : {
        "Model" : LogisticRegression(solver="liblinear", multi_class="auto"),
        'HyperParams' : {
            "C" : [1,5,10],
            "max_iter" : [100, 500, 1000]
        }
    },
    
    "GaussinaNB" : {
        "Model" : GaussianNB(priors=None, var_smoothing=1e-09),
        "HyperParams" : {}
    },
    
    "MultinominalNB" : {
        "Model" : MultinomialNB(alpha=1.0, fit_prior=True, class_prior=None),
        "HyperParams" : {}
    },
    
    "DecisionTree": {
        "Model" : DecisionTreeClassifier(),
        "HyperParams" : {}
    } 
    
}

In [128]:
# empty list for scores
scores = []

# looping through all algorithms to find best model and hyperparams that has max accuracy
for algo, hyperparam in Methods.items():
    # building model
    clf = GridSearchCV(hyperparam["Model"], hyperparam["HyperParams"], cv = 4, return_train_score = False)
    # fitting model to dataset
    clf.fit(irisdata.data, irisdata.target)
    # appending modelname, score, params dictonary to scores
    scores.append({
        "Model" : algo,
        "BestScore" : clf.best_score_,
        "BestParams" : clf.best_params_
    })

In [129]:
# sorting scores based on accuracy
scores = sorted(scores, key = lambda x: x["BestScore"], reverse = True)

# create dataframe from scores
data = pd.DataFrame(scores, columns = ["Model", "BestScore", "BestParams"])

In [130]:
data

Model  BestScore  \
0                         SVM   0.980085   
1  RandomForestClassification   0.966750   
2                         KNN   0.966750   
3          LogisticRegression   0.966394   
4              MultinominalNB   0.966394   
5                DecisionTree   0.959993   
6                  GaussinaNB   0.953414   

                                        BestParams  
0  {'C': 5.0, 'gamma': 'auto', 'kernel': 'linear'}  
1                             {'n_estimators': 30}  
2                               {'n_neighbors': 5}  
3                       {'C': 10, 'max_iter': 100}  
4                                               {}  
5                                               {}  
6                                               {}

#### we can say best hyperparameters are c = 5.0, gamma = "auto", kernel = "linear" for SVM model with accuracy 98.01 %